enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+access to them even if the IPs associated with the corresponding DNS records change dynamically.
+* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
+allow/deny access to all the subdomains belonging to the wildcard DNS names.
+* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
+in EgressFirewall rules can be tracked.
+* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
+rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
+* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the
+new plugin enabled.
+
+### Goals
+
+* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall rule containing
+a DNS name if the IPs associated with it change dynamically.
+* Support usage of wildcard DNS names in EgressFirewall rules.
+* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current
+IPs and the corresponding TTL information.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin.
+* Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
+(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
+for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
+be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
+based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
+CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an
+OpenShift cluster adminisrator.
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the
+plugin will update the ``.status`` of the matching ``DNSName`` CR(s) with the DNS name along with the IPs and
+the corresponding TTL and the next lookup time based on the TTL. The OVN-K master will watch the ``DNSName``
+CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
+referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).
+
+OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed
```

Comment:
Yes. This the current behavior of OVN-K to trigger the lookups based on the TTL values. I have added it here as this behavior is still required for supporting DNS rules in EgressFirewall (both regular and wildcard).

Reply:
Currently in OVN-K both TTL and next lookup time is stored in data structure used for DNS names. If TTL is not used anywhere apart from calculating next lookup time then it can be removed.